src/compiler/crystal/program.cr

```diff
@@ -326,10 +326,18 @@ module Crystal
 
     # Returns the `Type` for `type | Nil`
     def nilable(type)
-      # Nil | Nil # => Nil
-      return self.nil if type == self.nil
-
-      union_of self.nil, type
+      case type
+      when self.nil
+        # Nil | Nil # => Nil
+        return self.nil
+      when UnionType
+        types = Array(Type).new(type.union_types.size)
+        types.concat type.union_types
+        types << self.nil
```

Comment:
Isn't `type.union_types` an `Array(Type)` already?
```suggestion
        types = type.union_types.dup
        types << self.nil
```

Reply:
That's what I did first, but then there's a potential realloc.

I really wish we had splatting inside array literal

Reply:
Ah, that's why. Then it should be `.size + 1` ?

Reply:
Oh, yeah, that. Thanks!